### PR TITLE
Custom Actions (get_relevant_areas action)

### PR DESCRIPTION
**Pull request by MaxGroenenboom**
Added support for custom actions and a custom action, `get_relevant_areas` which only returns build-able areas at this moment.

Travis: [Build Status](https://travis-ci.org/Denpeer/tygron/builds/136373549)

User story:
As a developer, I want my virtual human to be able to dynamically determine areas to build on.

### CODE REVIEW

**Comment by MaxGroenenboom** [comment]
To add a custom action:
Create a Class that implements `CustomAction` and add an instance of it to the constructor of `ActionContainer`.
To add a custom area type to `get_relevant_areas`:
Create a Class that implements `RelevantAreasAction` and add an instance of it to the constructor of `GetRelevantAreas`. A `RelevantAreasAction` can also be added to `ActionContainer` to make it possible to call it directly.

**Comment by nskartoredjo** [comment]
There are still some debug lines in the code. I suppose those could be removed?

**Comment by MaxGroenenboom** [comment]
@nskartoredjo I'll take a look at the code, I'm sure I removed most of the debug lines but I might've missed something.
Oh nvm I know which lines you mean, I suppose they can be removed now ^^

File: contextvh/src/main/java/contextvh/actions/GetRelevantAreasBuild.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+				} catch (TranslationException e) {
+					TLogger.exception(e);
+					continue;
+				}
+				numPolys++;
+				GetRelevantAreas.debug("Added " + mp.toText());
+			}
+		}
+		createdPercept.addParameter(results);
+		GetRelevantAreas.debug("created result");
```
Comment (nskartoredjo) [inline contextvh/src/main/java/contextvh/actions/GetRelevantAreasBuild.java]: debug line

File: contextvh/src/main/java/contextvh/actions/GetRelevantAreasBuild.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+		// Remove all pieces of land that cannot be build on (water).
+		GetRelevantAreas.debug("removing water");
+		constructableLand = MapUtils.removeWater(connectionID, constructableLand);
+
+		// Remove all pieces of reserved land.
+		GetRelevantAreas.debug("removing reserved");
+		constructableLand = MapUtils.removeReservedLand(connectionID, constructableLand);
+
+		// Remove all pieces of occupied land.
+		GetRelevantAreas.debug("removing buildings");
```
Comment (nskartoredjo) [inline contextvh/src/main/java/contextvh/actions/GetRelevantAreasBuild.java]: debug line

File: contextvh/src/main/java/contextvh/actions/GetRelevantAreasBuild.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+		final Integer stakeholderID = caller.getStakeholder().getID();
+		MultiPolygon constructableLand = MapUtils.getMyLands(connectionID, stakeholderID);
+
+		// Remove all pieces of land that cannot be build on (water).
+		GetRelevantAreas.debug("removing water");
+		constructableLand = MapUtils.removeWater(connectionID, constructableLand);
+
+		// Remove all pieces of reserved land.
+		GetRelevantAreas.debug("removing reserved");
```
Comment (nskartoredjo) [inline contextvh/src/main/java/contextvh/actions/GetRelevantAreasBuild.java]: debug line

File: contextvh/src/main/java/contextvh/actions/GetRelevantAreasBuild.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	protected MultiPolygon getUsableArea(final ContextEntity caller, final ParameterList parameters) {
+		// Get a MultiPolygon of all lands combined.
+		GetRelevantAreas.debug("combining land");
+		Integer connectionID = caller.getSlotConnection().getConnectionID();
+
+		final Integer stakeholderID = caller.getStakeholder().getID();
+		MultiPolygon constructableLand = MapUtils.getMyLands(connectionID, stakeholderID);
+
+		// Remove all pieces of land that cannot be build on (water).
+		GetRelevantAreas.debug("removing water");
```
Comment (nskartoredjo) [inline contextvh/src/main/java/contextvh/actions/GetRelevantAreasBuild.java]: debug line

File: contextvh/src/main/java/contextvh/actions/GetRelevantAreas.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+				new Numeral(mp.getArea())
+		);
+	}
+
+	/**
+	 * Print a debug message.
+	 * Is a separate function to make it easier to remove debug messages.
+	 * @param message The message.
+	 */
+	public static void debug(final String message) {
```
Comment (nskartoredjo) [inline contextvh/src/main/java/contextvh/actions/GetRelevantAreas.java]: debug method

File: contextvh/src/main/java/contextvh/actions/GetRelevantAreas.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	 * @param callID		The ID of the call.
+	 * @param parameters	A ParameterList of parameters provided by the agent.
+	 * @return The constructed Percept.
+	 * @throws TranslationException  When an invalid internal action parameter is provided.
+	 */
+	private Percept createPercept(final ContextEntity caller, final String actionType,
+			final Number callID, final ParameterList parameters) throws TranslationException {
+		Percept result = new Percept("relevant_areas");
+		result.addParameter(new Numeral(callID));
+		debug("action called");
```
Comment (nskartoredjo) [inline contextvh/src/main/java/contextvh/actions/GetRelevantAreas.java]: debug line

File: contextvh/src/main/java/contextvh/actions/GetRelevantAreas.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+		System.out.println(internalActions);
```
Comment (nskartoredjo) [inline contextvh/src/main/java/contextvh/actions/GetRelevantAreas.java]: print line

**Comment by nskartoredjo** [comment]
@MaxGroenenboom I made a short comment under every debug line I could find. Should save you some time

**Comment by nskartoredjo** [comment]
@MaxGroenenboom the custom action is now not able to run an internal action after completion if I am right?

**Comment by MaxGroenenboom** [comment]
I'm pretty sure all of the debug lines have been removed, I don't know what you mean exactly by "after completion"? When a method has finished executing it can't do anything anymore at all...

File: contextvh/src/main/java/contextvh/ContextEnv.java [hunk 8]
```diff
@@ -67,23 +70,23 @@ protected boolean isSupportedByType(final Action action, final String type) {
...
      * @return returns a {@link Percept} if the {@link Action} returns one.
      * @throws ActException thrown if the action failed to execute
      */
     @Override
-    protected Percept performEntityAction(final String entity,
+    protected Percept performEntityAction(final String ent,
                                           final Action action) throws ActException {
         try {
-            getEntity(entity).performAction(action);
+            getEntity(ent).performAction(action);
```
Comment (spiderbiggen) [inline contextvh/src/main/java/contextvh/ContextEnv.java]: Why did you change the variable name from entity to ent

File: contextvh/src/main/java/contextvh/ContextEnv.java [hunk 9]
```diff
@@ -107,7 +110,8 @@ protected Percept performEntityAction(final String entity,
      */
     @Override
     public TygronEntity createNewEntity(final EntityListener listener, final String stakeholder, final Integer slot) {
-        return new ContextEntity(listener, stakeholder, slot);
+    	entity =  new ContextEntity(listener, stakeholder, slot);
```
Comment (spiderbiggen) [inline contextvh/src/main/java/contextvh/ContextEnv.java]: In the case of our agents this is fine but what happens if we make a mas file with two stakeholders in it?

**Comment by nskartoredjo** [comment]
@MaxGroenenboom It is not possible to have an action COS_ACTION(par1, par2, ...) that does its stuff, and at the end uses the internal action INT_ACTION(par1, par2, ...) (where the custom action delivers the needed parameters for the internal action)?

If I think of a custom action I would think that after I performed the custom action in the GOAL agent, it performed an action, and not only create a percept.

File: contextvh/src/main/java/contextvh/util/CoordinateUtils.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	/**
+	 * Divide a <code>Coordinate</code> with a number.
+	 * @param c The coordinate.
+	 * @param num The number.
+	 * @return A new Coordinate.
+	 */
+	public static Coordinate divide(final Coordinate c, final double num) {
+		return new Coordinate(c.x / num, c.y / num, c.z / num);
```
Comment (spiderbiggen) [inline contextvh/src/main/java/contextvh/util/CoordinateUtils.java:55]: maybe check if num is 0 here?

File: contextvh/src/main/java/contextvh/util/MapUtils.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+		return getZonesCombined(connectionID, zoneIDs);
+	}
+
+	/**
+	 * Returns a Geometry of all land of the stakeholder with ID stakeholderID.
+	 * @param connectionID The id of the connection.
+	 * @param stakeholderID The ID of the stakeholder.
+	 * @return The resulting Geometry.
+	 */
+	public static MultiPolygon getMyLands(final Integer connectionID, final Integer stakeholderID) {
```
Comment (spiderbiggen) [inline contextvh/src/main/java/contextvh/util/MapUtils.java]: maybe call this getStakeholderLands as it is not actually specific to any stakeholder

File: contextvh/src/test/java/contextvh/TestEnvironment.java [hunk 12]
```diff
@@ -11,31 +11,44 @@
...
+/**
+ * Tests the additional functionality in ContextEnv.
+ * @author Max Groenenboom
+ */
+public class TestEnvironment extends tygronenv.TestEnvironment {
 
-	private EisEnv env;
-    private static final String STAKEHOLDERS = "stakeholders";
+	private ContextEnv env;
+	private static final String STAKEHOLDERS = "stakeholders";
```
Comment (spiderbiggen) [inline contextvh/src/test/java/contextvh/TestEnvironment.java:21]: why are these tabs and not 4 spaces?

File: contextvh/src/test/java/contextvh/actions/GetRelevantAreasBuildTest.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+ *
+ */
+public class GetRelevantAreasBuildTest {
+
+	private static final String STAKEHOLDERS = "stakeholders";
+	private static final String MUNICIPALITY = "MUNICIPALITY";
+	private static final String PROJECT = "project";
+	private static final Identifier PROJECTNAME = new Identifier("testutilsmap");
+
+	private static final double AREA_MUNICIPALITY = (1000 - 0) * (1000 - 0);
```
Comment (spiderbiggen) [inline contextvh/src/test/java/contextvh/actions/GetRelevantAreasBuildTest.java:36]: why did you put - 0 in there? also you could probably just make it 1_000_000

File: contextvh/src/main/java/contextvh/ContextEnv.java (hunk 8 above)
Comment (MaxGroenenboom) [inline contextvh/src/main/java/contextvh/ContextEnv.java]: Because of checkstyle (it would hide a field).

File: contextvh/src/main/java/contextvh/util/CoordinateUtils.java (hunk 10 above)
Comment (MaxGroenenboom) [inline contextvh/src/main/java/contextvh/util/CoordinateUtils.java:55]: I think one should be able to multiply by 0 :P

File: contextvh/src/main/java/contextvh/util/MapUtils.java (hunk 11 above)
Comment (MaxGroenenboom) [inline contextvh/src/main/java/contextvh/util/MapUtils.java]: Good point, will do that right away.

File: contextvh/src/test/java/contextvh/TestEnvironment.java (hunk 12 above)
Comment (MaxGroenenboom) [inline contextvh/src/test/java/contextvh/TestEnvironment.java:21]: Because the connector uses tabs?

File: contextvh/src/test/java/contextvh/actions/GetRelevantAreasBuildTest.java (hunk 13 above)
Comment (MaxGroenenboom) [inline contextvh/src/test/java/contextvh/actions/GetRelevantAreasBuildTest.java:36]: It's `1000 * 1000` because it is an area of 1000*1000. I don't know about the `- 0`'s, I didn't write that...

File: contextvh/src/test/java/contextvh/TestEnvironment.java (hunk 12 above)
Comment (spiderbiggen) [inline contextvh/src/test/java/contextvh/TestEnvironment.java:21]: ah I see now nevermind

File: contextvh/src/main/java/contextvh/util/CoordinateUtils.java (hunk 10 above)
Comment (spiderbiggen) [inline contextvh/src/main/java/contextvh/util/CoordinateUtils.java:55]: I meant for the division but yeah multiplying by 0 should be possible

File: contextvh/src/main/java/contextvh/ContextEnv.java (hunk 8 above)
Comment (spiderbiggen) [inline contextvh/src/main/java/contextvh/ContextEnv.java]: ah yes of course you added an entity field

File: contextvh/src/main/java/contextvh/util/CoordinateUtils.java (hunk 10 above)
Comment (MaxGroenenboom) [inline contextvh/src/main/java/contextvh/util/CoordinateUtils.java:55]: Oh right. Because they are doubles I think it actually is possible to divide by zero? You'll get `infinity`'s though. I don't really see a reason to catch that.

**Comment by tvleest** [comment]
Looks like I'm not getting back any percept.

**Comment by MaxGroenenboom** [comment]
@tvleest Yeah, I found the culprit, testing now, if it works I'll push.

**Comment by MaxGroenenboom** [comment]
@tvleest That should be fixed now, it needs some more testing though.

**Comment by tvleest** [comment]
Works for me now.

File: contextvh/src/test/java/contextvh/actions/GetRelevantAreasBuildTest.java (hunk 13 above)
Comment (klikooo) [inline contextvh/src/test/java/contextvh/actions/GetRelevantAreasBuildTest.java:36]: It was to show that its land starts at (0,0) and ends at (1000,1000)

File: contextvh/src/main/java/contextvh/ContextEnv.java (hunk 9 above)
Comment (klikooo) [inline contextvh/src/main/java/contextvh/ContextEnv.java]: Yes, you are right! We will fix this in a few minutes.

File: contextvh/src/main/java/contextvh/util/CoordinateUtils.java (hunk 10 above)
Comment (spiderbiggen) [inline contextvh/src/main/java/contextvh/util/CoordinateUtils.java:55]: Alright was just a suggestion anyway

**Comment by MichielvdBerg** [comment]
It works for the us, We approve. \RAMpestampers.

**Comment by lisv** [comment]
CodeFox approves this as well. The code looks fine and works for us. Also we really need this feature.

**Comment by nskartoredjo** [comment]
Could you perhaps raise the version number?

**Comment by MaxGroenenboom** [comment]
@nskartoredjo We could do that, but is it really necessary?

**Comment by nskartoredjo** [comment]
It seems appropriate after each major change. Anyway it works with our agent so we approve. -CTD

**Comment by levilime** [comment]
Our agent is still functional with these new changes, therefore Danshal agrees with the merge :+1:
